### PR TITLE
Hpcc 15612 Embedded Couchbase Plugin

### PR DESCRIPTION
**Pull request by rpastrana**

### CODE REVIEW

**Comment by rpastrana** [comment]
@jamienoss please review

**Comment by rpastrana** [comment]
Not squashing the multiple commits for history keeping. To review, use "Files changed" tab.

**Comment by rpastrana** [comment]
@xwang2713 we will need to install libcouchbase on our build machines.

**Comment by rpastrana** [comment]
@dcamper please review.

File: cmake_modules/FindCOUCHBASE.cmake [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  IF (NOT ("${osdir}" STREQUAL "unknown"))
+    FIND_PATH(
+      LIBCOUCHBASE_INCLUDE_DIR
+      NAMES couchbase.h
+      PATHS "${EXTERNALS_DIRECTORY}/couchbase/${osdir}/include" "${EXTERNALS_DIRECTORY}/couchbase/include"
+      PATH_SUFFIXES include libcouchbase
+      NO_DEFAULT_PATH
+    )
+
+    FIND_LIBRARY(
```
Comment (dcamper) [inline cmake_modules/FindCOUCHBASE.cmake]: Is it correct to be finding MySQL libraries here?

File: plugins/couchbase/couchbaseembed.hpp [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    {
+        int count = 0;
+        const RtlFieldInfo * const *fields = record->queryFields();
+        assertex(fields);
+        while (*fields++)
+            count++;
+        return count;
+    }
+
+    void handleDeserializeOutcome(DeserializationResult resultcode, const char * targetype, const char * culpritvalue)
```
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.hpp]: Can this be static?

File: plugins/couchbase/couchbaseembed.hpp [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    private:
+        StringBuffer m_connectionString;
+        Couchbase::Client * m_pCouchbaseClient;
+        Couchbase::Status  m_connectionStatus;
+        Couchbase::Query  * m_pQuery;
+
+        CouchbaseConnection(const CouchbaseConnection &);
+    };
+
+    class CouchbaseRowBuilder : public CInterfaceOf<IFieldSource>
```
Comment (dcamper) [inline plugins/couchbase/couchbaseembed.hpp]: The code for this class feels like it should be included in the .cpp file instead of the header.  Was there a dependency issue?

**Comment by dcamper** [comment]
@rpastrana  A couple of minor notes and questions.  Overall it looks great.

File: plugins/couchbase/couchbaseembed.hpp (hunk 2 above)
Comment (rpastrana) [inline plugins/couchbase/couchbaseembed.hpp]: Yes, and I will make it such...

**Comment by rpastrana** [comment]
@dcamper please review changes based on your comments:  		29397b9

**Comment by rpastrana** [comment]
@Michael-Gardner please review build logic

**Comment by rpastrana** [comment]
@richardkchapman please review.

File: .gitmodules [hunk 4]
```diff
@@ -37,3 +37,6 @@
 [submodule "system/lz4_sm/lz4"]
     path = system/lz4_sm/lz4
     url = https://github.com/hpcc-systems/lz4.git
+[submodule "plugins/couchbase/libcouchbase-cxx"]
+	path = plugins/couchbase/libcouchbase-cxx
+	url = https://github.com/couchbaselabs/libcouchbase-cxx.git
```
Comment (richardkchapman) [inline .gitmodules]: We should fork the couchbaselabs repo into hpcc-systems, and use our fork. Otherwise we would break should they decide to rename/move/delete/rebase their repo.

Comment (richardkchapman) [inline .gitmodules]: (I have done the fork - can you update your PR)

**Comment by richardkchapman** [comment]
@rpastrana Looks good - one comment

**Comment by jamienoss** [comment]
@richardkchapman & @rpastrana am I still wanted to look over this?

File: system/jlib/jptree.cpp [hunk 5]
```diff
@@ -1057,7 +1057,7 @@ void PTree::setName(const char *_name)
         name = NULL;
     else
     {
-        if (!validateXMLTag(_name))
```
Comment (rpastrana) [inline system/jlib/jptree.cpp]: This unrelated change snuck in, and can't revert since the extra space doesn't pass our check-in criteria

**Comment by rpastrana** [comment]
@jamienoss, another set of eyes looking over this wouldn't hurt, but we are looking forward to closing this out soon. Let me know if you can take a look soon, otherwise I'll ask Richard to merge. Thanks.

**Comment by jamienoss** [comment]
@rpastrana Sure thing, I'll take a look over it today.

File: system/jlib/jptree.cpp (hunk 5 above)
Comment (richardkchapman) [inline system/jlib/jptree.cpp]: No big deal, it can stay

Comment (jamienoss) [inline system/jlib/jptree.cpp]: @rpastrana You can use "git commit -ns" to bypass this.

**Comment by jamienoss** [comment]
@rpastrana, @richardkchapman all looks good to me
